File: cpp/src/solutions/letter_combinations_of_a_phone_number.cpp
// LeetCode 17. Letter Combinations of a Phone Number in C++

#include "common.h"

class Solution {
 public:
  vector<string> letterCombinations(string digits) {
    const string letters[10] = {"",    "",    "abc",  "def", "ghi",
                                "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> ans;
    if (digits.size() == 0) {
      return ans;
    }
    ans = {""};
    for (auto digit : digits) {
      vector<string> tmp;
      auto& letter = letters[digit - '0'];
      for (auto& str : ans) {
        for (auto ch : letter) {
          tmp.push_back(str + ch);
        }
      }
      ans = tmp;
    }
    return ans;
  }
};

TEST(letterCombinations, example1) {
  string digits = "23";
  vector<string> ans = {"ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"};
  EXPECT_EQ(Solution().letterCombinations(digits), ans);
}

TEST(letterCombinations, example2) {
  string digits = "";
  vector<string> ans;
  EXPECT_EQ(Solution().letterCombinations(digits), ans);
}

TEST(letterCombinations, example3) {
  string digits = "2";
  vector<string> ans = {"a", "b", "c"};
  EXPECT_EQ(Solution().letterCombinations(digits), ans);
}

File: rust/src/solutions/letter_combinations_of_a_phone_number.rs
// LeetCode 17. Letter Combinations of a Phone Number in Rust

use super::Solution;

impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        let letters = vec![
            "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz",
        ];
        let mut ans = Vec::new();
        if digits.len() == 0 {
            return ans;
        }
        ans.push(String::from(""));
        for digit in digits.as_bytes() {
            let mut tmp = Vec::new();
            let letter = letters[(digit - b'0') as usize];
            for string in &ans {
                for ch in letter.chars() {
                    tmp.push(string.to_owned() + &ch.to_string());
                }
            }
            ans = tmp;
        }
        ans
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let digits = String::from("23");
        let ans: Vec<String> = vec![
            "ad".to_string(),
            "ae".to_string(),
            "af".to_string(),
            "bd".to_string(),
            "be".to_string(),
            "bf".to_string(),
            "cd".to_string(),
            "ce".to_string(),
            "cf".to_string(),
        ];
        assert_eq!(Solution::letter_combinations(digits), ans);
    }

    #[test]
    fn example2() {
        let digits = String::from("");
        let ans: Vec<String> = vec![];
        assert_eq!(Solution::letter_combinations(digits), ans);
    }

    #[test]
    fn example3() {
        let digits = String::from("2");
        let ans: Vec<String> = vec!["a".to_string(), "b".to_string(), "c".to_string()];
        assert_eq!(Solution::letter_combinations(digits), ans);
    }
}
